### DISCUSSION TITLE
FastAPI with sync ORM using threading.local

### DISCUSSION THREAD

**Discussion by ghost** [fastapi/fastapi#10996] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI
from pydal import DAL, Field

db = DAL("CONNECTION_STRING")

db.define_table("user",
    Field('first_name', length=128),
    Field('last_name', length=128),
)

app = FastAPI()

@app.post("/")
def root():
    db(db.auth_user.id == 1).update(first_name="hello")
    db.commit()
        
    return 'done'
```

### Description

I am in the process of migrating Web2Py code to FastAPI and I would like to leave all the sync pyDAL database code as-is. Upon doing more research I've found conflicting/confusing statements regarding the usage of sync/non-async code. I only want to use sync code. I've seen multiple GitHub issues that devs using sync orms such peewee or Django ORM run into connection pool issues. 

FastAPI docs state that use sync endpoints can be if you have sync code which will run an external threadpool. Does the external threadpool work with threading.local? [pyDAL](https://github.com/web2py/pydal/) uses threading.local, and I can't seem to find a good answer if everything will work as expected if I only use sync. 

A comment from tiangolo https://github.com/tiangolo/fastapi/issues/2619#issuecomment-1359044461 says that using thread.local variables is unsafe, but I'm not sure if that's specific to async or not.

[pyDAL docs](https://py4web.com/_documentation/static/en/chapter-07.html) and [https://github.com/web2py/pydal/issues/427](url) and https://groups.google.com/g/web2py/c/7Dl1lUeotgk that it is thread safe as long as the global variables are only being read from, but also that it writes to thread.local.

Any ideas if this is possible or if I have misunderstood anything so far? Lastly, is this worth the effort? I'm set on using pyDAL would it be a better idea to use a WSGI framework such as Flask or Falcon?

Thanks!

### Operating System

Windows

### FastAPI Version

0.109.0

### Pydantic Version

2.5.3

### Python Version

3.11.4

**Comment by acidjunk** [comment]
A web2py: I also had some projects with that :)

You're probably fine by just not using async functions/endpoints.
It might be a good idea to test the threading, it probably also depends on what DB you use; thread safe should be enough.

You could also start extra workers with just one thread: to achieve process level isolation.

**Reply by ghost** [reply]
It's a very interesting framework to say the least!

How would I test the threading, some kind of load tests, multiple long open sync requests?
How can I start extra workers with one thread? Is that the same as using the ThreadPoolExecutor like this https://github.com/tiangolo/fastapi/issues/4221#issuecomment-984724324

Thanks!
